### ISSUE TITLE
Add object destructuring support for sort-vars

### ISSUE THREAD

**Issue by remcohaszing** [eslint/eslint#11078]
**What rule do you want to change?**

`sort-vars`

**Does this change cause the rule to produce more or fewer warnings?**

More

**How will the change be implemented? (New option, new default behavior, etc.)?**

This will force assignment through object destructuring to be sorted alphabetically. This will also affect destructed objects as function argument.

A new boolean option will be added: `destructuring`. This will default to `false` for backwards compatibility.

**Please provide some example code that this change will affect:**

```js
const { b, a } = foo;  // `a` should be sorted before `b`.

// This pattern is common in React components.
const Component = (props) => {
  const { id, className } = props;  // `className` should be sorted before id`.
  // etc…
};

const OtherComponent = ({ id, className }) => {  // `className` should be sorted before id`.
  // etc…
};

// This also affects nested properties
const { d, c: { b, a } } = foo;  // `a should be sorted before `b`. `c` should be sorted before `d`.
```

**What does the rule currently do for this code?**

Nothing

**What will the rule do after it's changed?**

It will require destructured properties to be sorted alphabetically, respecting the existing `ignoreCase` flag. So `a` should ge before `b` in the first example.

**Are you willing to submit a pull request to implement this change?**

Yes

**Comment by kaicataldo** [comment]
I think it makes sense to have a rule that does this. One confusing thing about the current rules is that we have at least one rule that has the word `object` in its name but ends up also checking imports and destructured assignments (example [here](https://eslint.org/docs/rules/object-curly-newline)), and given that we have the [`sort-keys`](https://eslint.org/docs/rules/sort-keys) rule it's not clear if this should then be in `sort-vars`, `sort-keys`, or a new rule entirely.

I'd love to have a way to clarify these rules so that it's clear by the name what they actually check.

**Comment by remcohaszing** [comment]
In addition, I think it should be allowed to violate the sorting if it’s needed to use one of the destructured variables is used as the default of another destructured variable.

E.g.

```js
const {
  c,
  a = c,
  b,
} = foo;
```

This does leave the question: _should `b` be sorted before `c` or after `a`?_

While typing this I figured it should probably be placed before `c`, as `a = c` is the code that makes the situation exceptional.
